File: credit.c
#include<stdio.h>
#include<cs50.h>
#include<math.h>
int noofdi(long x);
int retdi(long y);
int sumoth(long z);
void typecard(long a);
long n;
int main(void)
{
    do
    {
        n=get_long("Number: ");
    }
    while(n<1);
    typecard(n);
    }
int noofdi(long x)
{   int r=0;
    while(x>=10)
    {
        x=x/10;
        r=r+1;
    }
   r=r+1; 
   return r;
}
int retdi(long y)
{    int sum=0;
     int m;
     for(int j=1;j<noofdi(n);j+=2)
     {    y=n;
         for(int i=0;i<j;i++)
     {
         y=y/10;
     }
     if (y>10)
     {
         m=(y%10)*2;
     }
     else 
     {
         m=y*2;
     }
     if (m<10)
     {
         sum+=m;
     } 
     else 
     {
         sum+=m/10+m%10;
     }
     } return sum;
}
int sumoth(long z)
{
    int sum1=0;
    for(int k=0;k<noofdi(n);k+=2)
     {  z=n;
        for(int j=0;j<k;j++)
        {
            z=z/10;
        }
        if (z>10)
        {
            sum1+=z%10;
               }
        else 
        {
            sum1+=z;
        }       
     } return sum1;
}
void typecard(long a)
{     
    int u=sumoth(n)+retdi(n);
         if (u%10==0)
          {  
              if (noofdi(n)==15)
              {   for(int i=13;i>0;i--)
                  {
                      n=n/10;
                  } 
                  if (n==34 || n==37)
                  {
                      printf("AMEX\n");
                      }
                  else 
                  {
                      printf("INVALID\n");
                  }    
              }
                else if (noofdi(n)==16)
              {   for(int i=14;i>0;i--)
                  {
                       n=n/10;
                }
                  if (n==51 || n==52 || n==53 || n==54 || n==55)
                  {
                      printf("MASTERCARD\n");
                           }
                  else if (n/10==4)
                  {
                      printf("VISA\n");
                           }  
                  else 
                  {
                      printf("INVALID\n");
                             }    
              }
              else if (noofdi(n)==13)
               {   for(int i=12;i>0;i--)
                   { 
                        n=n/10;
                         }
                  if (n==4)      
                  {
                      printf("VISA\n");
                          }
                   else
                   {
                         printf("INVALID\n");
                         }    
                 }
            }
      else
      {
          printf("INVALID\n");
      }
   }

File: credit/credit.py
import re
from cs50 import get_int, get_string

def main():
    n = get_int("NUMBER:")
    s = get_string("number:")
    if check(n) == False:
        print("INVALID ")
    else:
        type(n,s)


def noofdig(n):
    count = 0
    while n >= 10:
        n = n / 10
        count += 1

    return count + 1


def check(n):
    sum1 = 0
    sum2 = 0
    x = noofdig(n)
    for i in range(0,x,2):
        y = n
        for j in range(i + 1):
            y = y // 10
        m = y % 10
        if m * 2 >= 10:
            sum1 += (m * 2) // 10 + (m * 2) % 10
        else:
            sum1 += m*2

    for k in range(0,x,2):
        y = n
        for l in range(0,k):
            y = y // 10
        m = y % 10
        sum2 += m

    if (sum1 + sum2) % 10 == 0:
        return True
    else:
        return False

def type(n,s):
    if noofdig(n) == 15:
        if re.search("^34(.*)$",s) or re.search("^37(.*)$",s):
            print("AMEX\n")
    elif noofdig(n) == 16:
        if re.search("^5(.*)$",s):
            print("MASTERCARD\n")
        elif re.search("^4(.*)$",s):
            print("VISA\n")
        else:
            print("INVALID\n")
    elif noofdig(n) == 13:
        if re.search("^4(.*)$",s):
            print("VISA\n")
    else:
        print("INVALID\n")

main()
